### ISSUE TITLE
Distributed slack only works for BalanceType.PROPORTIONAL_TO_LOAD

### ISSUE THREAD

**Issue by tso-martin** [powsybl/pypowsybl#860]
### Describe the current behavior

Conducting a load flow with a distributed slack currently only works correctly for `BalanceType.PROPORTIONAL_TO_LOAD`.
I add examples for `PROPORTIONAL_TO_GENERATION_P_MAX`, `PROPORTIONAL_TO_GENERATION_P`, and `PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR` that are currently not working as expected.

### Describe the expected behavior

Depending on the chosen `BalanceType` the power of generators or loads affected by the slack should change.

### Describe the steps

```python
#%%
import pypowsybl as pp
import pypowsybl.network as pn
import pypowsybl.loadflow as lf
from pypowsybl.sensitivity import create_dc_analysis
from pypowsybl._pypowsybl import BalanceType

# %% Initial Power Flow
network = pp.network.create_ieee30()
results = lf.run_ac(network)
load_p_init = network.get_loads().p
gen_p_init = network.get_generators().p

#%% Increase Power of Generator B5-G
network.update_generators(id='B5-G', target_p=100)

#%% PROPORTIONAL_TO_LOAD works
params = lf.Parameters(
    read_slack_bus=False, 
    write_slack_bus=False, 
    distributed_slack=True, 
    balance_type=BalanceType.PROPORTIONAL_TO_LOAD)
lf.run_ac(network, params) # distributed_active_power=-99.15416476433062
load_p = network.get_loads().p # different from load_p_init

#%% PROPORTIONAL_TO_GENERATION_P_MAX does not work
params = lf.Parameters(
    read_slack_bus=False, 
    write_slack_bus=False, 
    distributed_slack=True, 
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_P_MAX)
lf.run_ac(network, params) # distributed_active_power=0.0
gen_p = network.get_generators().p # All generators at p_target

#%% PROPORTIONAL_TO_GENERATION_P does not work
params = lf.Parameters(
    read_slack_bus=False, 
    write_slack_bus=False, 
    distributed_slack=True, 
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_P)
lf.run_ac(network, params) # distributed_active_power=0.0
gen_p = network.get_generators().p # All generators at p_target

#%% Slack bus selection by OpenLoadFlow does not work
params = lf.Parameters(
    read_slack_bus=False, 
    write_slack_bus=False, 
    distributed_slack=True, 
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_P,
    provider_parameters={'maxSlackBusCount': '3', 'slackBusSelectionMode': 'NAME', 'slackBusesIds': ['VL1_0', 'VL2_0', 'VL5_0']})
lf.run_ac(network, params) # --> TypeError related to slackBusesIds

#%% PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR does not work
network.create_extensions('activePowerControl', pd.DataFrame.from_records(index='id', data=[
        {'id': 'B1-G', 'participate': True, 'participation_factor': 0.2},
        {'id': 'B2-G', 'participate': True, 'participation_factor': 0.4},
        {'id': 'B5-G', 'participate': True, 'participation_factor': 0.4}
    ]))
params = lf.Parameters(
    read_slack_bus=False, 
    write_slack_bus=False, 
    distributed_slack=True, 
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR)
lf.run_ac(network, params) # distributed_active_power=0.0
```

### Environment

PyPowSybl 1.7.0
Windows 10

**Comment by jeandemanged** [comment]
Hi @tso-martin ,

Thank you for your feedback.

- The IEEE test cases do not define Pmax for all generators, and open load flow has a rule to exclude generators with unreasonable Pmax from slack distribution. See [OLF configuration page](https://powsybl.readthedocs.io/projects/powsybl-open-loadflow/en/latest/loadflow/parameters.html), parameter `plausibleActivePowerLimit`.
- about `slackBusIds` parameters, the array has to be provided as a string, separate the IDs with a comma or a semicolon at your convenience. This is a documentation issue we will have to resolve [here](https://powsybl.readthedocs.io/projects/pypowsybl/en/stable/user_guide/loadflow.html#parameters-specific-to-a-provider)

For more details here adjusted test script with my comments marked `# jd /` where slack distribution would be okay.

```python
#%%
import pypowsybl as pp
import pypowsybl.network as pn
import pypowsybl.loadflow as lf
from pypowsybl.sensitivity import create_dc_analysis
from pypowsybl._pypowsybl import BalanceType

import pandas as pd

# jd / activate logging for troubleshooting, see https://powsybl.readthedocs.io/projects/pypowsybl/en/stable/user_guide/logging.html
import logging
logging.getLogger('powsybl').setLevel(logging.INFO)
logging.basicConfig()

# %% Initial Power Flow
network = pp.network.create_ieee30()
results = lf.run_ac(network)
load_p_init = network.get_loads().p
gen_p_init = network.get_generators().p

#%% Increase Power of Generator B5-G
network.update_generators(id='B5-G', target_p=100)

#%% PROPORTIONAL_TO_LOAD works
params = lf.Parameters(
    read_slack_bus=False,
    write_slack_bus=False,
    distributed_slack=True,
    balance_type=BalanceType.PROPORTIONAL_TO_LOAD)
lf.run_ac(network, params) # distributed_active_power=-99.15416476433062
load_p = network.get_loads().p # different from load_p_init

#%% PROPORTIONAL_TO_GENERATION_P_MAX does not work
# jd / In logs:
#      WARNING:powsybl:Network {CC0 SC0}: 6 generators have been discarded from active power control because of maxP not plausible
#      WARNING:powsybl:Failed to distribute slack bus active power mismatch, -97.93076579530535 MW remains
#      See https://powsybl.readthedocs.io/projects/powsybl-open-loadflow/en/latest/loadflow/parameters.html
#      Parameter plausibleActivePowerLimit
#      The IEEE test cases have 9999 MW Pmax for all generators.
#      Two choices: either modify plausibleActivePowerLimit, or update the generators Pmax
params = lf.Parameters(
    read_slack_bus=False,
    write_slack_bus=False,
    distributed_slack=True,
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_P_MAX,
    provider_parameters={'plausibleActivePowerLimit': '10000'})
lf.run_ac(network, params) # distributed_active_power=0.0
gen_p = network.get_generators().p # All generators at p_target

#%% PROPORTIONAL_TO_GENERATION_P does not work
# jd / Same issue as PROPORTIONAL_TO_GENERATION_P_MAX
params = lf.Parameters(
    read_slack_bus=False,
    write_slack_bus=False,
    distributed_slack=True,
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_P,
    provider_parameters={'plausibleActivePowerLimit': '10000'}) # jd / added
lf.run_ac(network, params) # distributed_active_power=0.0
gen_p = network.get_generators().p # All generators at p_target

#%% Slack bus selection by OpenLoadFlow does not work
# jd / thanks for reporting, documentation issue here on https://powsybl.readthedocs.io/projects/pypowsybl/en/stable/user_guide/loadflow.html#parameters-specific-to-a-provider
#      parameters of type string lists should be provided as a simple string and not an array, but the documentation is not clear enough, will improve this.
params = lf.Parameters(
    read_slack_bus=False,
    write_slack_bus=False,
    distributed_slack=True,
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_P,
    provider_parameters={'maxSlackBusCount': '3', 'slackBusSelectionMode': 'NAME', 'slackBusesIds': 'VL1_0,VL2_0,VL5_0', 'plausibleActivePowerLimit': '10000'})  # jd / modified
lf.run_ac(network, params) # --> TypeError related to slackBusesIds

#%% PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR does not work
# jd / Same issue as PROPORTIONAL_TO_GENERATION_P_MAX
network.create_extensions('activePowerControl', pd.DataFrame.from_records(index='id', data=[
    {'id': 'B1-G', 'participate': True, 'participation_factor': 0.2},
    {'id': 'B2-G', 'participate': True, 'participation_factor': 0.4},
    {'id': 'B5-G', 'participate': True, 'participation_factor': 0.4}
]))
params = lf.Parameters(
    read_slack_bus=False,
    write_slack_bus=False,
    distributed_slack=True,
    balance_type=BalanceType.PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR,
    provider_parameters={'plausibleActivePowerLimit': '10000'}) # jd / added
lf.run_ac(network, params) # distributed_active_power=0.0
```

**Comment by jeandemanged** [comment]
Documentation fix to be done about STRING_LIST parameters (e.g. open-loadflow `slackBusesIds`)

**Comment by tso-martin** [comment]
Dear @jeandemanged,
Thank you for your help. Now it works.

**Comment by jeandemanged** [comment]
> Dear @jeandemanged, Thank you for your help. Now it works.

Dear @tso-martin,
Your welcome, I did the documentation fix here: #861.
Do not hesitate to provide feedback either here on GitHub, or you may also join the PowSyBl Slack channel (link at the top of the README.md) where you could discuss with other PyPowSyBl developers, maintainers, and users.
